pubgrep/pubgrep.py

```diff
@@ -0,0 +1,540 @@
+"""
+
+import sys
+from urllib.parse import quote
+import argparse
+from pathlib import Path
+import subprocess as sp
+import shutil
+
+import requests
+
+try:
+    from version import __version__  # type: ignore # pylint: disable=import-error
+except ImportError:
+    __version__ = "0.0.0"
+
+
+def header() -> str:
+    """
+    This function is used to print the header of the program.
+    """
+    headerstr = (
+        "---------------------------------------------------------------------\n"
+        f"                          PubGrep {__version__}\n"
+        "- This Program tries to search CIDs from the Pubchem Database based -\n"
+        "- on a list of compounds given as Input. Afterwards it creates sdf  -\n"
+        "-   Files for each Compound given in an appropriate subdirectory.   -\n"
+        "-     If you are using this program extensively (like, a lot!)      -\n"
+        "-   for your Research, please consider citing 10.1039/D3RA01705B    -\n"
+        "-                          MS, MM, 2021-2024                        -\n"
+        "---------------------------------------------------------------------\n\n"
+    )
+    return headerstr
+
+
+# Custom Exceptions for xTB and structure conversions
+class XtbFailure(Exception):
+    "Raised when the xTB calculation does not provide the expected output."
+
+
+### URL request functions
+def rawurlencode(string):
+    """
+    This function is used to encode the string to be used in the URL.
+    """
+    return quote(string, safe="-_.~a-zA-Z0-9")
+
+
+def test_pubchem_server(verbosity):
+    """
+    This function is used to test the connection to the PubChem server.
+    """
+    response = requests.get(
+        "https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/cid/1/cids/TXT", timeout=10
+    )
+    if response.status_code == 200 and response.text.strip() == "1":
+        if verbosity > 1:
+            print("PubChem Server is working fine.")
+    else:
+        raise ConnectionError(
+            "No connection could be established. Check if you have access to the internet."
+        )
+
+
+def search_compound(compound, input_format):
+    """
+    This function is used to search the compound in the PubChem database.
+    """
+    base_url = "https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/"
+    search_url = ""
+    if input_format == "inchi":
+        # Prepare the data payload
+        data = {"inchi": compound}
+        search_url = f"{base_url}inchi/{rawurlencode(compound)}/cids/TXT"
+    elif input_format == "name":
+        search_url = f"{base_url}name/{rawurlencode(compound)}/cids/TXT"
+    elif input_format == "cid":
+        search_url = f"{base_url}cid/{compound}/property/IUPACname/TXT"
+    elif input_format in ["smile", "smiles"]:
+        search_url = f"{base_url}smiles/{rawurlencode(compound)}/cids/TXT"
+    elif input_format in ["cas", "regid"]:
+        search_url = f"{base_url}xref/RegistryID/{rawurlencode(compound)}/cids/TXT"
+
```

Comment:
Nitpick: Including an else block that just throws an error for unrecognized formats, helps avoiding wrong input types.